### PR TITLE
Support gateway packet compression

### LINKED ISSUE #10: Support gateway packet compression

### PR DESCRIPTION
**Pull request by ChallahuAkbar**
This _should_ make it so that gateway packet compression is handled correctly. Closes #10.

### CODE REVIEW

File: src/discordcr/websocket.cr [hunk 1]
```diff
@@ -33,6 +34,16 @@ module Discord
       )
     end
 
+    def on_binary(&handler : Packet ->)
+      @websocket.on_binary do |binary|
+        message = IO::Memory.new(binary)
+        Zlib::Reader.open(message, sync_close = true) do |reader|
```
[the remark this answers is not in the document]
Comment (ChallahuAkbar) [reply src/discordcr/websocket.cr]: _note to self: actually read what I type. . ._

But yeah, true enough.

File: src/discordcr/websocket.cr [hunk 2]
```diff
@@ -33,6 +34,16 @@ module Discord
       )
     end
 
+    def on_binary(&handler : Packet ->)
+      @websocket.on_binary do |binary|
+        message = IO::Memory.new(binary)
```
[the remark this answers is not in the document]
Comment (ChallahuAkbar) [reply src/discordcr/websocket.cr]: I used `message` to mirror the behavior of `on_message`, but `io` makes sense.

**Comment by meew0** [comment]
The problem is that the `d` object of packets can be one of a large variety of objects, so it's not easily possible to use standard JSON parsing methods for this.

File: src/discordcr/websocket.cr [hunk 3]
```diff
@@ -33,10 +61,21 @@ module Discord
       )
     end
 
+    def on_binary(&handler : Packet ->)
+      @websocket.on_binary do |binary|
+        io = IO::Memory.new(binary)
+        Zlib::Reader.open(io) do |reader|
+          payload = Packet.from_json(reader)
+          @logger.debug "[WS IN] (compressed, #{binary.size}) #{payload.to_json}"
```
Comment (z64) [inline src/discordcr/websocket.cr]: This needs to be followed by `if @logger.debug?` (see its sister line in `on_message`)

**Pull request merged**